Tensile/Source/client/source/PerformanceReporter.cpp

```diff
@@ -0,0 +1,178 @@
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ *******************************************************************************/
+
+#include "PerformanceReporter.hpp"
+
+#include <cmath>
+#include <cstddef>
+#include <limits>
+#include <string>
+#include <unordered_map>
+
+namespace Tensile
+{
+    namespace Client
+    {
+        std::shared_ptr<PerformanceReporter> PerformanceReporter::Default(po::variables_map const& args)
+        {
+            int     deviceIndex = args["device-idx"].as<int>();
+            double  l2ReadHits = args["perf-l2-read-hits"].as<double>();
+            double  l2WriteHits = args["perf-l2-write-hits"].as<double>();
+            double  readEff = args["perf-read-efficiency"].as<double>();
+            bool    mfma = args["perf-mfma"].as<bool>();
+ 
+            return std::make_shared<PerformanceReporter>(deviceIndex, l2ReadHits, l2WriteHits, readEff, mfma);
+        }
+        
+        PerformanceReporter::PerformanceReporter(int deviceIndex, double l2ReadHits, double l2WriteHits, double readEff, bool mfma)
+        {
+            hipGetDeviceProperties(&m_props, deviceIndex);
+            setNumCUs();
+            setMemoryBusWidth();
+            setPerfModel(l2ReadHits, l2WriteHits, readEff, mfma);
+            setMagicNum();
+            m_deviceProps = true;
+            
+            perf.l2ReadHitRate = getL2ReadHits();
+            perf.l2WriteHitRate = getL2WriteHits();
+            perf.readEff = getReadEff();
+            perf.CUs = getNumCUs();
+        }
+        
+        void PerformanceReporter::reportValue_uint(std::string const& key, uint64_t value) 
+        {
+            if(key == ResultKey::SpeedGFlops && m_deviceProps) 
+            {
+                reportValue_numeric(key, value);
+            }
+        }
+
+        void PerformanceReporter::reportValue_double(std::string const& key, double value) 
+        {
+            if(key == ResultKey::ClockRateSys && m_deviceProps)
+            {
+                m_clockMhz = value;
+                perf.clock = getClock();
+            }
+            if(!std::isnan(m_clockMhz) && m_deviceProps)
+            {
+                m_peakGFlops = getNumCUs()*getMagicNum()*getReadMultiplier()*m_clockMhz/1000;
+                perf.peakGFlops = getPeakGFlops();
+            }
+            if(key == ResultKey::ClockRateMem && m_deviceProps)
+            {
+                m_memClockMhz = value;
+                perf.memClock = getMemClock();
+                m_memBandwidthMBps = m_memoryBusWidth*m_memClockMhz;
+                perf.memBandwidthMBps = getMemBandwidthMBps();
+            }
+            if(key == ResultKey::SpeedGFlops && m_deviceProps) 
+            {
+                reportValue_numeric(key, value);
+            }
+        }
+
+        template <typename T> 
+        void PerformanceReporter::reportValue_numeric(std::string const& key, T value)
+        {
```

Comment:
Most of the code should be in this unified function.  The other `reportValue` functions should **just** call this one.